test/haxesharp/collections/DictionaryTest.hx

```diff
@@ -0,0 +1,253 @@
+package haxesharp.collections;
+
+using haxesharp.collections.Dictionary;
+using haxesharp.test.Assert;
+using haxesharp.exceptions.ArgumentException;
+using haxesharp.exceptions.ArgumentNullException;
+using haxesharp.exceptions.KeyNotFoundException;
+
+class DictionaryTest
+{
+    @Test
+    public function dictionaryAllowsIntKeys()
+    {
+        testDictionaryGetSetAndRemove(1, 1, 7);
+        testDictionaryMapConstructor([
+            1 => 1,
+            2 => 2,
+            3 => 3
+        ]);
+    }
+
+    @Test
+    public function dictionaryAllowsStringKeys()
+    {
+        testDictionaryGetSetAndRemove("foo", "bar", "key that doesn't exist");
+        testDictionaryMapConstructor([
+            "1" => 1,
+            "2" => 2,
+            "3" => 3
+        ]);
+    }
+
+    @Test
+    public function dictionaryAllowsEnumValueKeys()
+    {
+        testDictionaryGetSetAndRemove(TestEnum.OneValue, 1, TestEnum.TwoValue);
+        testDictionaryMapConstructor([
+            TestEnum.OneValue => 1,
+            TestEnum.TwoValue => 2,
+            TestEnum.ThreeValue => 3
+        ]);
+    }
+
+    @Test
+    public function dictionaryAllowsObjectKeys()
+    {
+        // dictionary uses reference equality
+        testDictionaryGetSetAndRemove(new TestObject(), 1, new TestObject());
+        var obj1 = new TestObject();
+        var obj2 = new TestObject();
+        var obj3 = new TestObject();
+        testDictionaryMapConstructor([
+            obj1 => 1,
+            obj2 => 2,
+            obj3 => 3
+        ]);
+    }
+
+    @Test
+    public function getThrowsKeyNotFoundExceptionWhenKeyDoesNotExist()
+    {
+        var testKeyNotFoundDictionary = new Dictionary<Int,Int>();
+        testKeyNotFoundDictionary.set(1, 1);
+
+        try
+        {
+            testKeyNotFoundDictionary.get(2);
+            Assert.fail("testKeyNotFoundDictionary.get(2) did not throw!");
+        }
+        catch (e:KeyNotFoundException)
+        {
+            //Pass
+        }
+    }
+
+    @Test
+    public function addThrowsArgumentExceptionOnDuplicateKey()
```

Comment:
I was going to say "are you missing a test where set throws on duplicate keys" but I realized that add vs. set have different semantics. I'm not sure if this is a Bad Thing.

But, this is a normal byproduct of our array accessors. This is fine.

Reply:
My idea was the duplicate the functionality of .NET Dictionary's `[]` setter, which will not throw on duplicate keys.